File: python/535. Encode and Decode TinyURL.py
class Codec:
    def __init__(self):
        self.cache = {}
        self.rev_cache = {}
        self.idx = 0
        
    def base64(self, num):
        a = []
        while num:
            a.append(num % 62)
            num /= 62
        a.reverse()
        b = []
        for i in range(len(a)):
            if a[i]<26:
                ch = chr(ord('a') + a[i])
            elif a[i]<52:
                ch = chr(ord('A') + a[i]-26)
            else:
                ch = chr(ord('0') + a[i]-52)
            b.append(ch)
        return ''.join(b)

    def encode(self, longUrl):
        """Encodes a URL to a shortened URL.
        
        :type longUrl: str
        :rtype: str
        """
        if not self.cache.has_key(longUrl):
            self.cache[longUrl] = self.base64(self.idx)
            self.rev_cache[self.cache[longUrl]] = longUrl
            self.idx += 1
        return self.cache[longUrl]


    def decode(self, shortUrl):
        """Decodes a shortened URL to its original URL.
        
        :type shortUrl: str
        :rtype: str
        """
        if shortUrl in self.rev_cache:
            return self.rev_cache[shortUrl]
        return ''
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(url))

File: go/src/leetcode/535/535. Encode and Decode TinyURL.go
/**
@author ZhengHao Lou
Date    2022/06/29
*/
package main

import (
	"encoding/base64"
	"fmt"
)

/**
https://leetcode.cn/problems/encode-and-decode-tinyurl/
*/
type Codec struct {
	i     int
	cache map[string]string
}

func Constructor() Codec {
	return Codec{
		cache: map[string]string{},
	}
}

// Encodes a URL to a shortened URL.
func (this *Codec) encode(longUrl string) string {
	e := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%v", this.i)))
	this.i++
	this.cache[e] = longUrl
	return e
}

// Decodes a shortened URL to its original URL.
func (this *Codec) decode(shortUrl string) string {
	return this.cache[shortUrl]
}

/**
 * Your Codec object will be instantiated and called as such:
 * obj := Constructor();
 * url := obj.encode(longUrl);
 * ans := obj.decode(url);
 */
